src/unstructured_client/_hooks/custom/split_pdf_hook.py

```diff
@@ -270,40 +334,47 @@ def before_request(
         if split_size >= page_count and page_count == len(pdf.pages):
             return request
 
-        pages = pdf_utils.get_pdf_pages(pdf, split_size=split_size, page_start=page_range_start, page_end=page_range_end)
-
-        # Use a variable to adjust the httpx client timeout, or default to 30 minutes
-        # When we're able to reuse the SDK to make these calls, we can remove this var
-        # The SDK timeout will be controlled by parameter
-        client_timeout_minutes = 60
-        if timeout_var := os.getenv("UNSTRUCTURED_CLIENT_TIMEOUT_MINUTES"):
-            client_timeout_minutes = int(timeout_var)
-
-        async def call_api_partial(page):
-            client_timeout = httpx.Timeout(60 * client_timeout_minutes)
-
-            async with httpx.AsyncClient(timeout=client_timeout) as client:
-                response = await request_utils.call_api_async(
-                    client=client,
-                    original_request=request,
-                    form_data=form_data,
-                    filename=file.file_name,
-                    page=page,
-                    limiter=limiter,
-                )
-
-                return response
+        if self.cache_tmp_data_feature:
+            pdf_chunk_paths = self._get_pdf_chunk_paths(
+                pdf,
+                operation_id=operation_id,
+                split_size=split_size,
+                page_start=page_range_start,
+                page_end=page_range_end
+            )
+            # force free PDF object memory
+            del pdf
+            pdf_chunks = self._get_pdf_chunk_files(pdf_chunk_paths)
```

Comment:
Question in general about the new process: how do we improve memory management in general?

So for example, if I have a 5000 page PDF that consumes ~50MB of memory per chunk (let's say this process breaks it down into 100 batches of 50 pages), I understand that when batch 1 of 50mb, it will write that to a file, batch 2 write to file and so on...


so if the job takes 20 minutes to complete, the overall memory usage stays low while we write to files.

However, at the end of a 20 minute time-frame once the last (50th) batch finally completes, we then take all of the 100 batches written in each file, and bring everything back into memory at this point in time to return it to the caller, right?  Are we optimizing purely for the duration of the run to keep memory low, but not concerned that, at the end the memory needs to go back up to resolve the file?

Reply:
Yes, you're right - we don't want to occupy the memory in a long process (and with hi res it can be tens or hundreds of minutes). But we cannot avoid loading the json files back to the memory because the interface must work the same way as for non `split-pdf` calls - so returning a Response with a json attached. That's why we won't avoid the spike at the end.

 Maybe we could replace the response `read` operation with something smart that would load the files one by one but I don't think we'll get much here.

Also - what's important in these changes - we no longer pass the bytes or io.BytesIO object to the request but a file object which is read to the socket in a buffered way.